### PR TITLE
NEW Update rulesets

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/.github/issues/198

### CODE REVIEW

File: funcs_utils.php [hunk 1]
```diff
@@ -152,7 +152,8 @@ function github_api($url, $data = [], $httpMethod = '')
 {
     // silverstripe-themes has a kind of weird redirect only for api requests
     $url = str_replace('/silverstripe-themes/silverstripe-simple', '/silverstripe/silverstripe-simple', $url);
-    info("Making curl request to $url");
+    $method = $httpMethod ? strtoupper($httpMethod) : 'GET';
+    info("Making $method curl request to $url");
```
Comment (emteknetnz) [inline funcs_utils.php:156]: This is an unrelated enhancement to make the console log a little nicer to see what HTTP method is being used

File: rulesets_command.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        // Get any additional branches to add
+        // Assumption is that if the default branch is main/master, then the repo uses
+        // a non-numeric style branching system (e.g. main, master) and that needs to be protected
+        // [0-9]* branch protection will still be applied, on the chance that the repo is converted
+        // to uses a numeric style branch system in the future and we would want branch protection
+        // to start immediately on the new branches
+        $additionalBranchConditions = [];
+        $defaultBranch = github_api("https://api.github.com/repos/$account/$repo")['default_branch'];
+        if (in_array($defaultBranch, ['main', 'master'])) {
+            $additionalBranchConditions[] = "refs/heads/$defaultBranch";
```
Comment (GuySartorelli) [inline rulesets_command.php:61]: Is there any harm in using `~DEFAULT_BRANCH` here? Or just always using `~ALL` regardless of what the default branch is?
I found those as documented special values in https://docs.github.com/en/rest/repos/rules?apiVersion=2022-11-28#create-a-repository-ruleset
> Also accepts `~DEFAULT_BRANCH` to include the default branch or `~ALL` to include all branches.

File: rulesets_command.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Update rulesets
+    foreach ($modules as $module) {
+        $account = $module['account'];
+        $repo = $module['repo'];
+
+        // Fetch existing rulesets
+        $rulesets = github_api("https://api.github.com/repos/$account/$repo/rulesets");
+        $branchRulesetID = 0;
+        $tagRulesetID = 0;
```
Comment (GuySartorelli) [inline rulesets_command.php:40]: Are we certain `0` isn't a valid ID? Might be safer to use `null` here, and explicitly check if the value is `null` below instead of falsy.

File: rulesets_command.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            $url = "https://api.github.com/repos/$account/$repo/rulesets/$branchRulesetID";
+            $apiCalls[] = [$url, $branchRuleset, 'PUT'];
+        }
+        if ($tagRulesetID) {
+            $url = "https://api.github.com/repos/$account/$repo/rulesets/$tagRulesetID";
+            $apiCalls[] = [$url, $tagRuleset, 'PUT'];
+        }
+
+        if ($input->getOption('dry-run')) {
+            info('Not updating rulesets on GitHub because --dry-run option is set');
```
Comment (GuySartorelli) [inline rulesets_command.php:98]: May be worth dumping `$apiCalls` either to a file or to the terminal here so we can see in a dry-run if the calls look like we expect them to.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: funcs_utils.php [hunk 5]
```diff
@@ -244,6 +245,44 @@ function output_repos_with_labels_updated()
...
+{
+    $ruleset = file_get_contents("rulesets/$type-ruleset.json");
+    if (!$ruleset) {
+        error("Could not read ruleset for $type");
+    }
+    $json = json_decode($ruleset, true);
+    if ($type == 'branch') {
+        $json['name'] = BRANCH_RULESET_NAME;
+    } elseif ($type === 'tag') {
+        $json['name'] = TAG_RULESET_NAME;
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply funcs_utils.php:276]: Seems more solid to just use a constant rather then risk not updating the json file

File: rulesets_command.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    $modules = [];
+    $modulesCurrentMajor = filtered_modules(MetaData::HIGHEST_STABLE_CMS_MAJOR, $input);
+    $modulesPreviousMajor = filtered_modules(MetaData::HIGHEST_STABLE_CMS_MAJOR - 1, $input);
+    foreach ([$modulesCurrentMajor, $modulesPreviousMajor] as $modulesList) {
+        foreach ($modulesList as $module) {
+            // Important! Only include modules on the "silverstripe" account
+            if ($module['account'] !== 'silverstripe') {
+                continue;
+            }
+            $modules[] = $module;
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply rulesets_command.php:28]: Updated

File: rulesets_command.php (hunk 3 above)
Comment (emteknetnz) [reply rulesets_command.php:40]: Updated

File: rulesets_command.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        }
+    }
+
+    // Update rulesets
+    foreach ($modules as $module) {
+        $account = $module['account'];
+        $repo = $module['repo'];
+
+        // Fetch existing rulesets
+        $rulesets = github_api("https://api.github.com/repos/$account/$repo/rulesets");
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply rulesets_command.php:38]: Updated

File: rulesets_command.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+        // Get any additional branches to add
+        // Assumption is that if the default branch is main/master, then the repo uses
+        // a non-numeric style branching system (e.g. main, master) and that needs to be protected
+        // [0-9]* branch protection will still be applied, on the chance that the repo is converted
+        // to uses a numeric style branch system in the future and we would want branch protection
+        // to start immediately on the new branches
+        $additionalBranchConditions = [];
+        $defaultBranch = github_api("https://api.github.com/repos/$account/$repo")['default_branch'];
+        if (in_array($defaultBranch, ['main', 'master'])) {
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply rulesets_command.php:60]: I think it's correct to hardcode. `main` is the current convention for non-numeric branches and `master` is the old convention

File: rulesets_command.php (hunk 2 above)
Comment (emteknetnz) [reply rulesets_command.php:61]: Leave as main/master.  We already have loads of `master` regular branch protections in places e.g. https://github.com/silverstripe/cow/settings/branches - if we introduce `~DEFAULT_BRANCH` we've now got an additional way to do the same thing which is just adds additional mental overhead for no difference in practical functionality

File: rulesets_command.php [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            $additionalBranchConditions[] = "refs/heads/$defaultBranch";
+        }
+
+        // Create rulesets
+        // Note: This will read from the "rulesets" directory
+        // In each of those json rulesets there is "bypass_actors"."actor_id" = 5
+        // This translates to the "Repository admin" role
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply rulesets_command.php:67]: I didn't find one. Instead I simply manually tested the different bypass levels to see if the github-action user could bypass or not

Note you get these by default + some teams linked to the repo

image

"Deal with that when it happens" is fine IMO, using this PR we can very easily change the bypass to maintainer if desired

File: rulesets_command.php (hunk 4 above)
Comment (emteknetnz) [reply rulesets_command.php:98]: Done

File: run.php [hunk 10]
```diff
@@ -19,13 +20,16 @@
 const TOOL_URL = 'https://github.com/silverstripe/module-standardiser';
 const PR_TITLE = 'MNT Run module-standardiser';
 const PR_DESCRIPTION = 'This pull-request was created automatically by [module-standardiser](' . TOOL_URL . ')';
+const BRANCH_RULESET_NAME = 'Silverstripe branch ruleset';
+const TAG_RULESET_NAME = 'Silverstripe tag ruleset';
```
[the remark this answers is not in the document]
Comment (emteknetnz) [reply run.php:24]: Done

File: funcs_utils.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply funcs_utils.php:276]: That's why I suggested adding a comment if that's a concern.
But not worth arguing about. Leave as is.

File: rulesets_command.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply rulesets_command.php:62]: I don't think it's correct to hardcode basically any time ever - but not worth arguing about. Leave as is.

File: rulesets_command.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply rulesets_command.php:69]: > I didn't find one. Instead I simply manually tested the different bypass levels to see if the github-action user could bypass or not

Gotta love when APIs require magic values but don't document those values lol.

> "Deal with that when it happens" is fine IMO, using this PR we can very easily change the bypass to maintainer if desired

My concern isn't in how difficult it would be to change later. My concern is that if we're forced to change it later we're effectively escalating the permissions for roles we don't want to escalate the permissions for. Community peer reviewers are probably gonna be at the maintainer level, I'd think.

File: rulesets_command.php (hunk 2 above)
Comment (GuySartorelli) [reply rulesets_command.php:63]: What about just `~ALL` regardless of what the default branch is? It seems like that simplifies things.

**Comment by GuySartorelli** [comment]
Just a couple of things above to consider. Up to you if you think changes are warranted or not.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Sorry, I thought I asked for this last time but it looks like I only thought about asking it: Please update the readme.

File: rulesets_command.php (hunk 2 above)
Comment (emteknetnz) [reply rulesets_command.php:63]: Means that it'll cover things like dependabot PRs, update-js PRs etc, which shouldn't be protected

Comment (GuySartorelli) [reply rulesets_command.php:63]: Ahh, I thought update-js PRs were done on a fork but I've just checked and they're not. Which is a little gross but it is what it is.

File: rulesets_command.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (emteknetnz) [reply rulesets_command.php:69]: Seems like we're then basically forced to make any new CMS squad members "Repository Admin" in that case so that they can merge PRs normally

Yeah that's probably not ideal

Having a look at https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization - what's there than would neccesitate 'maintain' over 'admin' for community peer reviewers?

Seems like it might be better to make the reviewers "Write" and then set the bypass on the rulesets to "Maintain" instead of "Repository admin"

[the remark this answers is not in the document]
Comment (GuySartorelli) [reply rulesets_command.php:69]: Oh nice, didn't know that matrix existed.
Yeah I forgot about "write" access lol. That will be perfect for reviewers. No worries then, leave as is. I'll make sure the card for creating the reviewer role explicitly gives them write access only.

**Comment by emteknetnz** [comment]
Have updated readme

**Review by GuySartorelli — APPROVED** [review]
LGTM
